scipy/special/xsf/zeta.h

```diff
@@ -1,10 +1,279 @@
+	std::complex<double> log_poch = std::log(z);
+	std::complex<double> log_factor = std::log(b) - std::log(N);
+	for (std::size_t i = 0; i < m; i++) {
+	    std::complex<double> term = std::exp(zeta_em_log_coeff(i) + log_factor + log_poch);
+	    result += term;
+	    if (std::abs(term)/std::abs(result) <= std::numeric_limits<double>::epsilon()) {
+		return result;
+	    }
+	    log_poch += std::log(z + static_cast<double>(2*i + 1)) + std::log(z + static_cast<double>(2*i + 2));
+	    log_factor -= 2*std::log(N);
+	}
+	// Euler-maclaurin absolute error estimate.
+	double error;
+	error = std::abs(std::exp(zeta_em_log_coeff(m) + log_factor + log_poch));
+	error *= std::abs((z + 2.0*m + 1.0)/(z.real() + 2.0*m + 1.0));
+	// convert to relative error estimate
+	error /= std::abs(result);
+	if (error > 1e-8) {
+	    if (error > 1e-1) {
+		/* If error estimate predicts we don't even get 1 digit of precision, return NaN
+		 * and signal no result */
+		set_error("zeta", SF_ERROR_NO_RESULT, NULL);
+		return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+	    }
+	    // Signal reduced precision.
+	    set_error("zeta", SF_ERROR_LOSS, NULL);
+	}
+	return result;
+    }
+
+    constexpr double zeta_borwein_coeff[] = {
+	1.0555078361382878e-38,
+	5.278594688527578e-35,
+	4.4014687322044963e-32,
+	1.467453546497519e-29,
+	2.617862196688831e-27,
+	2.900097799958025e-25,
+	2.184440361492933e-23,
+	1.1890977312913296e-21,
+	4.8871396166872276e-20,
+	1.5672253698802734e-18,
+	4.022931234264572e-17,
+	8.435973533351745e-16,
+	1.469296379914116e-14,
+	2.1548747054571902e-13,
+	2.6919530537535124e-12,
+	2.8925409162768484e-11,
+	2.6957505693699856e-10,
+	2.194772239130839e-09,
+	1.57078229370057e-08,
+	9.936187220749133e-08,
+	5.581721578217702e-07,
+	2.796271112037765e-06,
+	1.253886254275813e-05,
+	5.049261002939051e-05,
+	0.00018312884459703666,
+	0.0005997690328552426,
+	0.0017780501082460968,
+	0.004781802283665968,
+	0.011690432287131671,
+	0.026034302929535964,
+	0.052922982472754856,
+	0.09842471411648858,
+	0.16789610796540344,
+	0.26350429194768626,
+	0.3819442810600665,
+	0.5137731385068898,
+	0.645292538541866,
+	0.7625449322050362,
+	0.8556063057019102,
+	0.921056062886525,
+	0.9616100580028147,
+	0.9835905431606953,
+	0.9939187229336753,
+	0.9980782525166648,
+	0.9994930141459889,
+	0.999891478844585,
+	0.9999819091990203,
+	0.9999977981288319,
+	0.9999998260580315,
+	0.9999999933099243
+    };
+
+    /* Compute riemann_zeta for complex input z using Borwein's algorithm. */
+    XSF_HOST_DEVICE inline std::complex<double> zeta_borwein(std::complex<double> z) {
+	std::complex<double> s1 = 0.0;
+	// Sum in reverse order because smaller terms come later.
+	for (int k = 49; k >= 0; k--) {
+	    double sign = std::pow(-1.0, k);
+	    std::complex<double> den = std::pow(k + 1, z);
+	    auto t1 = sign * (zeta_borwein_coeff[k] - 1.0) / den;
+	    s1 += t1;
+	}
+	std::complex<double> result = s1;
```

Comment:
Not a big deal, just curious - why not just use `result = 0` from the start? Or if the intent was to keep the sum and coefficient conceptually separate, I guess I would not have expected `result` to be modified in place; rather, the coefficient would be calculated separately and then `return coefficient * sum`.

Reply:
I think it's just a vestige of this being written in a different way earlier, with it getting left in an in-between state which still works. I'll change it.